File: packages/shared/src/utils/tag-extractor.ts
/**
 * Tag Extraction Utilities
 *
 * Extracts hashtags from note content.
 * Pattern: #tagname (must start with a letter, followed by letters, numbers, or underscores)
 */

import { getUrlRanges, type UrlRange } from './web-link-utils';
import { isValidHeadingId } from './heading-extractor';

/**
 * Maximum allowed length for a tag name (excluding the # prefix)
 * Tags longer than this will be truncated
 */
export const MAX_TAG_LENGTH = 50;

/**
 * Regex pattern for matching hashtags
 * Pattern: # followed by a letter, then zero or more letters/numbers/underscores
 */
export const HASHTAG_PATTERN = /#[a-zA-Z][a-zA-Z0-9_]*/g;

/**
 * Check if a position falls within any URL range
 */
function isPositionInUrl(position: number, urlRanges: UrlRange[]): boolean {
  for (const range of urlRanges) {
    if (position >= range.start && position < range.end) {
      return true;
    }
  }
  return false;
}

/**
 * Extract all hashtags from text
 *
 * Hashtags that appear within URLs (as fragments or query params) are NOT extracted.
 * For example, in "https://example.com/page#section", the "#section" is a URL fragment
 * and will not be returned as a tag.
 *
 * @param text Plain text content
 * @returns Array of unique tag names (lowercase, without # prefix)
 */
export function extractTags(text: string): string[] {
  // Input validation
  if (text === null || text === undefined || typeof text !== 'string') {
    return [];
  }

  // Get all URL ranges first so we can skip hashtags inside URLs
  const urlRanges = getUrlRanges(text);

  const tags = new Set<string>();
  const regex = new RegExp(HASHTAG_PATTERN.source, HASHTAG_PATTERN.flags);

  let match;
  while ((match = regex.exec(text)) !== null) {
    const matchPosition = match.index;

    // Skip this hashtag if it's inside a URL
    if (isPositionInUrl(matchPosition, urlRanges)) {
      continue;
    }

    // Get the tag without # prefix
    const tagWithoutHash = match[0].slice(1);

    // Skip if this looks like a heading ID (h_XXXXXXXX format)
    // These appear in same-note heading links like [[#h_abc12xyz]]
    if (isValidHeadingId(tagWithoutHash)) {
      continue;
    }

    // Convert to lowercase
    let tag = tagWithoutHash.toLowerCase();

    // Enforce max length
    if (tag.length > MAX_TAG_LENGTH) {
      tag = tag.slice(0, MAX_TAG_LENGTH);
    }

    tags.add(tag);
  }

  return Array.from(tags);
}

File: packages/ios/Sources/Utilities/TagExtractor.swift
//
//  TagExtractor.swift
//  NoteCove
//
//  Created by NoteCove Contributors
//  Copyright © 2025 NoteCove Contributors. All rights reserved.
//

import Foundation

/// Utility for extracting hashtags from note content
///
/// Matches the tag extraction logic from desktop:
/// - Pattern: #[a-zA-Z][a-zA-Z0-9_]*
/// - Tags are case-insensitive (normalized to lowercase)
/// - Maximum tag length: 50 characters
public struct TagExtractor {
    /// Maximum length for a tag
    public static let maxTagLength = 50

    /// Regular expression pattern for matching hashtags
    /// Must start with # followed by a letter, then letters/numbers/underscores
    private static let hashtagPattern = #"#[a-zA-Z][a-zA-Z0-9_]*"#

    /// Extract all hashtags from text content
    ///
    /// - Parameter text: The text to extract tags from
    /// - Returns: Array of unique tag names (lowercase, without # prefix)
    public static func extractTags(from text: String) -> [String] {
        guard let regex = try? NSRegularExpression(pattern: hashtagPattern, options: []) else {
            print("[TagExtractor] Failed to create regex pattern")
            return []
        }

        let range = NSRange(text.startIndex..., in: text)
        let matches = regex.matches(in: text, options: [], range: range)

        var tags = Set<String>()

        for match in matches {
            guard let matchRange = Range(match.range, in: text) else {
                continue
            }

            var tag = String(text[matchRange])

            // Remove # prefix
            tag = String(tag.dropFirst())

            // Convert to lowercase
            tag = tag.lowercased()

            // Enforce max length
            if tag.count > maxTagLength {
                tag = String(tag.prefix(maxTagLength))
            }

            tags.insert(tag)
        }

        return Array(tags).sorted()
    }
}
